pkg/didcomm/crypto/jwe/authcrypt/encrypt.go

```diff
@@ -0,0 +1,392 @@
+	APU string `json:"apu,omitempty"`
+	IV  string `json:"iv,omitempty"`
+	Tag string `json:"tag,omitempty"`
+	KID string `json:"kid,omitempty"`
+	OID string `json:"oid,omitempty"`
+}
+
+// New will create an encrypter instance to 'AuthCrypt' payloads for the given sender and recipients arguments
+// and the encryption alg argument. Possible algorithms supported are:
+// C20P (chacha20-poly1035 ietf)
+// XC20P (xchacha20-poly1035 ietf)
+// The returned crypter contains all the information required to encrypt payloads.
+func New(sender keyPair, recipients []*[chacha20poly1305.KeySize]byte, alg ContentEncryption) (*Crypter, error) {
+	var nonceSize int
+	switch alg {
+	case C20P:
+		nonceSize = chacha20poly1305.NonceSize
+	case XC20P:
+		nonceSize = chacha20poly1305.NonceSizeX
+	default:
+		return nil, errors.New(fmt.Sprintf("encryption algorithm '%s' not supported", alg))
+	}
+	if len(recipients) == 0 {
+		return nil, errors.New("empty recipients keys, must have at least one recipient")
+	}
+	var recipientsKey []*[chacha20poly1305.KeySize]byte
+	recipientsKey = append(recipientsKey, recipients...)
+
+	c := &Crypter{
+		sender,
+		recipientsKey,
+		alg,
+		nonceSize,
+	}
+
+	if !isKeyPairValid(sender) {
+		return nil, errors.New(fmt.Sprintf("sender keyPair not supported, it must have %d bytes keys", chacha20poly1305.KeySize))
+	}
+
+	return c, nil
+}
+
+func isKeyPairValid(kp keyPair) bool {
+	if kp.priv == nil || kp.pub == nil {
+		return false
+	}
+
+	return true
+}
+
+// Encrypt will JWE encode the payload argument for the sender and recipients
+// Using (X)Chacha20 encryption algorithm and Poly1035 authenticator
+func (c *Crypter) Encrypt(payload string) ([]byte, error) {
+	headers := jweHeaders{
+		"typ": "prs.hyperledger.aries-auth-message",
+		"alg": "ECDH-SS+" + string(c.alg) + "KW",
+		"enc": string(c.alg),
+	}
+
+	aad := c.buildAAD()
+	aadEncoded := base64.URLEncoding.EncodeToString(aad)
+
+	encHeaders, err := json.Marshal(headers)
+	if err != nil {
+		return nil, err
+	}
+	// build the Payload's AAD string
+	pldAad := base64.URLEncoding.EncodeToString(encHeaders) + "." + aadEncoded
+
+	// generate a new nonce for this encryption
+	nonce := make([]byte, c.nonceSize)
+	_, err = RandReader.Read(nonce)
+	if err != nil {
+		return nil, err
+	}
+	nonceEncoded := base64.URLEncoding.EncodeToString(nonce)
+
+	var cek *[chacha20poly1305.KeySize]byte
+
+	// generate a cek for encryption (it will be treated as a symmetric key)
+	_, cek, err = box.GenerateKey(RandReader)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a cipher for the given nonceSize and generated cek above
+	cipher, err := createCipher(c.nonceSize, cek[:])
+	if err != nil {
+		return nil, err
+	}
+
+	pld := []byte(payload)
+
+	// encrypt payload using generated nonce, payload and its AAD
+	// the output is a []byte containing the cipherText + tag
+	symOutput := cipher.Seal(nil, nonce, pld, []byte(pldAad))
+
+	// symOutput has a length of len(pld) + poly1035.TagSize
```

Comment:
I'd personally move this tag manipulation stuff to another method.
(to make things more obvious and less cluttered).

Reply:
moved both tag and ciphertext into separate functions as they're reused for recipient keys as well